pkg/agent/metrics/metrics.go

```diff
@@ -46,13 +47,34 @@ var (
 
 // AgentMetrics includes all the metrics of the proxy agent.
 type AgentMetrics struct {
-	latencies *prometheus.HistogramVec
-	failures  *prometheus.CounterVec
+	serverFailures *prometheus.CounterVec
+	dialFailures   *prometheus.CounterVec
+	dialLatencies  *prometheus.HistogramVec
+	connections    *prometheus.GaugeVec
+	packets        *prometheus.CounterVec
 }
 
 // newAgentMetrics create a new AgentMetrics, configured with default metric names.
 func newAgentMetrics() *AgentMetrics {
-	latencies := prometheus.NewHistogramVec(
+	serverFailures := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: namespace,
+			Subsystem: subsystem,
+			Name:      "server_connection_failure_count",
+			Help:      "Count of failures to send to or receive from the proxy server, labeled by the direction (from_server or to_server)",
+		},
+		[]string{"direction"},
```

Comment:
These should all be gRPC errors, right? It might be useful to add the error code as a label: https://pkg.go.dev/google.golang.org/grpc@v1.51.0/status#Code

Reply:
This was not a real diff, just a move.

We could deprecate this metric with a stream errors (receive + sent) count, let me try adding here.

A recent Recv error log I found does appear to have code: `"could not read stream" err="rpc error: code = Unavailable desc = error reading from server: read tcp 10.76.0.7:55908->10.128.0.66:8132: read: connection timed out"`

Reply:
It might be sufficient to just look up the logs in that case. I don't think it's worth adding a new metric for it.

Reply:
Latest snapshot has stream errors, PTAL